constraint/pkg/apis/templates/transform_test.go

```diff
@@ -0,0 +1,355 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+func TestAddPreserveUnknownFields(t *testing.T) {
+	trueBool := true
+	testCases := []struct {
+		name string
+		v    *apiextensionsv1beta1.JSONSchemaProps
+		exp  *apiextensionsv1beta1.JSONSchemaProps
+	}{
+		{
+			name: "nil properties",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Properties: nil,
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+				Properties:             nil,
+			},
+		},
+		{
+			name: "no information",
+			v:    &apiextensionsv1beta1.JSONSchemaProps{},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+			},
+		},
+		{
+			name: "type object",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Type: "object",
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+				Type:                   "object",
+			},
+		},
+		{
+			name: "type array",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Type: "array",
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+				Type:                   "array",
+			},
+		},
+		{
+			name: "map with empty JSONSchemaProps value",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+					"foo": {},
+				},
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+				Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
```

Comment:
Shall we also test if preserve-unknown is specifically set to false?

Reply:
Good catch.  Will do now